### DISCUSSION TITLE
[Bug] Service Overriding Troubles

### DISCUSSION THREAD

**Discussion by rroessler** [microsoft/monaco-editor#3842] [Q&A]
### Reproducible in vscode.dev or in VS Code Desktop?

- [X] Not reproducible in [vscode.dev](https://vscode.dev) or VS Code Desktop

### Reproducible in the monaco editor playground?

- [ ] Not reproducible in [the monaco editor playground](https://microsoft.github.io/monaco-editor/playground.html)

### Monaco Editor Playground Link

> This link showcases the inability to "delete" code.

https://microsoft.github.io/monaco-editor/playground.html?source=v0.36.1#XQAAAAJrAgAAAAAAAABBqQkHQ5NjdMjwa-jY7SIQ9S7DNlzs5W-mwj0fe1ZCDRFc9ws9XQE0SJE1jc2VKxhaLFIw9vEWSxW3yscw11tiW_1uhrpu6yp6HeMeFg4D87fsZ9XASQdfuW03940xAddmItzxv09CitTmGEgQHWuwJ-iF7TB0AGegdjyvVx9mCOssYcS9IqEfYXAbEXRJlXgh4INf7lx9YsGYYE6aFGKyY7VxLiW1i15QM9T9NihULPdO3tURA0gYVMea3Qlvxbx9G6NlHcjpvXZPaiYFMYEXG6DpCudhmXQufv7SUJSUQhAyUx2CFaz6En6gST19_1nZiVixAoRguqp2TFdhXDF0-uuiM403sr9gFD17kRBN4PlS47pJ0JXHI-YtISoy8EPCbYfHq5Yd3z3GNq9e48mVNqvnaheD3a4pKCLyeencswggXpaSrPDzLcPkJvTI71tMRmIgpft9kgzkJmuE_IMCBw

> This link showcases a monkey-patch way of overriding services failing due to obfuscation.

https://microsoft.github.io/monaco-editor/playground.html?source=v0.36.1#XQAAAAJfAQAAAAAAAABBqQkHQ5NjdMjwa-jY7SIQ9S7DNlzs5W-mwj0fe1ZCDRFc9ws9XQE0SJE1jc2VKxhaLFIw9vEWSxW3yscwzYWtRgmSatbGbs83zUMeFg4D87fsZ9XASQdfuW03940xAddmItzxv09CitTmGEgQHWuwJ-iF7TB0AGegdjyvVx9mCOssYcWRmIq46paEzB2eWmxJ77UkNboL7v-HaA4s1ktoE3s2mhbDOsRYMZe0fRgJNFXQN-jQHjd80TaPTc6W0uDkKRrOvEUfCuysgV93Mn5_9DDGcHKAbT4vKq9jk7pI899UtP5hEGLK6Mkz0JutztrFRn0ty9swLrtXEfLTSLYF1Ge0m5IlyVh0lelzevFHBZ__sRaM4A

### Monaco Editor Playground Code

```typescript
const options = { value: 'console.log("Hello, World")', language: 'javascript' };

// Creating this way results in "backspace" key not working.
const editor = monaco.editor.create(document.getElementById("container"), options, {
	// [A] below override might be poorly formed
	codeEditorService: {
		onCodeEditorAdd() {},
		onCodeEditorRemove() {},
		listCodeEditors() { return []; },

		onDiffEditorAdd() {},
		onDiffEditorRemove() {},
		listDiffEditors() { return []; },

		willCreateCodeEditor() {},

		addCodeEditor() {},

		getFocusedCodeEditor() { return editor; },
		openCodeEditor() { return editor; }
	}
});

// [B] The old monkey-patch way to override services now broken due to obfuscation.
editor._codeEditorService.openCodeEditor = function() { return editor; };
```

### Reproduction Steps

Start by removing section [B] to show that the delete key does not work. Then enable [B] and remove the property below [A]. This will show that due to obfuscation we cannot override code-editor service methods.

### Actual (Problematic) Behavior

Overriding the "codeEditorService" in the conventional way results in an editor where the "Backspace" key does not work. Funnily enough, the "deleteLeft" command for the editor still works as expected when triggered manually.

In the second case, since version 0.35.0 there is now obfuscation that prevents overriding individual methods on the "codeEditorService".

### Expected Behavior

Even though the override service is ill-formed, this should not result in the "Backspace" key not working.

### Additional Context

It would be great to have some method on the editor API to expose underlying services, or at the very least better documentation in regards to overriding them.

**Comment by hediet** [comment]
> since version 0.35.0 there is now obfuscation that prevents overriding individual methods on the "codeEditorService".

I just disabled that in https://github.com/microsoft/vscode/pull/177223
